### ISSUE TITLE
WarpX on new HPC cluster at UCI

### ISSUE THREAD

**Issue by floresv299** [ECP-WarpX/WarpX#3938]
Hello, I'm trying to install and run Warpx on my institution's cluster. I have some experience installing softwares but there are some dependancies that I'm not sure what they do or how to get them. Thank you for the help !

**Comment by ax3l** [comment]
Hi @floresv299,

Thank you for reaching out! We are glad to help you to get started.

The best way to run on institutional clusters is to use their modules for base dependencies, e.g., compilers, MPI and any dependency we can find, and only add what we need manually.

In general, all (potential and optional) dependencies of WarpX are listed here:
https://warpx.readthedocs.io/en/latest/install/dependencies.html#install-dependencies

I would recommend to start with the following, which you should get from `module load` on your HPC system:
- compiler
- CMake
- MPI

Let's then get into a few details of your cluster:
- Is there a public documentation to the machine you are referring to? Can you provide us with that link to guide you?
  - Is this a CPU or GPU machine?
- Do you already run other codes on the machine?
  - Are there job scripts that you already use?
- For WarpX, which features do you like to use?
  - GPU acceleration?
  - 1D/2D/3D and/or RZ geometry?
  - Advanced field solvers (PSATD)?

Based on this, we can guide you best how to do the next steps.

Just to share upfront, we document a lot of clusters around the world in our manual's HPC section. We can derive from one that is close to yours and even add your workflow there as well if you like.
https://warpx.readthedocs.io/en/latest/install/hpc.html

**Comment by ax3l** [comment]
Hi @floresv299, how is it going? :)

**Comment by floresv299** [comment]
Dear Axel,

Thank you very much for reaching out, apologies for the late reply. Here
are the answers to some of your questions:

1) Here is the link to the documentation for our HPC
https://rcic.uci.edu/hpc3/index.html
This machine is able to run on a CPU and GPU.

2) I have used both CPU and GPU computing for EPOCH and most recently
for FBPIC. Below, I will attach some of the job scripts.

3) As for features it would be nice to have GPU acceleration and 1/2/3D
geometry. In terms of advanced field solvers, what would you recommend? The
work that our lab does is LWFA and solid-target interaction,
utilizing boosted frames to look at betatron emission and high harmonic
generation.

Thank you again for the help, it means a lot !

**Comment by ax3l** [comment]
Hi @floresv299 ,

Thanks for the details! Your cluster looks like a Linux-based, Module-provided, Slurm-scheduled system, so we can build instructions similar to [Lawrencium (LBNL)](https://warpx.readthedocs.io/en/latest/install/hpc/lawrencium.html), adjusting exact module names and versions as needed.

The job scripts did not attach, unfortunately. Can you post them in the GitHub issue?

For features that sounds like we will just do a full-flavored install, including pseudo-spectral field solvers and all geometries :)

**Comment by ax3l** [comment]
> there are some dependancies that I'm not sure what they do or how to get them

So getting started on this, we need to find the software we need in
- https://rcic.uci.edu/hpc3/software-tutorial.html
- https://rcic.uci.edu/hpc3/modules-update.html

Since I have no access to that specific system, can you save the output of `module avail` in a text file and post it here? That way, we can pick what we like and compile the extras we might need.

**Comment by floresv299** [comment]
Hi Axel, 

apologies, I did not notice that the documents were not posted. Full-flavored install sounds great thank you!

[HPC3_FBPIC_batch.pdf](https://github.com/ECP-WarpX/WarpX/files/11751049/HPC3_FBPIC_batch.pdf)
[HPC3_EPOCH_job.pdf](https://github.com/ECP-WarpX/WarpX/files/11751050/HPC3_EPOCH_job.pdf)
[Warpx_on_HPC3_UCI.pdf](https://github.com/ECP-WarpX/WarpX/files/11751053/Warpx_on_HPC3_UCI.pdf)

**Comment by ax3l** [comment]
Awesome, thanks a lot for the details.

So, I think for a GPU build, module-wise we will need:
```bash
# required dependencies
module load cmake/3.22.1
module load gcc/11.2.0
module load cuda/11.7.1
module load openmpi/4.1.2/gcc.11.2.0

# optional: for QED support with detailed tables
module load boost/1.78.0/gcc.11.2.0

# optional: for openPMD and PSATD+RZ support
module load hdf5/1.14.0/gcc.11.2.0-openmpi.4.1.2

# optional: CCache
#module load ccache  # missing

# optional: for Python bindings
module load python/3.10.2

# optimize CUDA compilation for V100
export AMREX_CUDA_ARCH=7.0

# compiler environment hints
export CXX=$(which g++)
export CC=$(which gcc)
export FC=$(which gfortran)
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=${CXX}
```

This will already build you **all of WarpX**, besides the RZ geometry with PSATD field solvers.

The only things we need to build (we can write a script) are:
- c-blosc (for compressed I/O)
- ADIOS2 (for faster I/O than HDF5)
- BLAS++ & LAPACK++ for RZ Geometry with PSATD field solver

Do you see an "OpenBLAS" or "LAPACK" module somewhere in `module av` once you have loaded the modules listed above?

**Comment by floresv299** [comment]
Thank you ! 

Yes I see both of them: lapack/3.9.0 and OpenBLAS/0.3.6 , OpenBLAS/0.3.19 , OpenBLAS/0.3.21

<img width="1190" alt="Screenshot 2023-06-14 at 3 55 23 PM" src="https://github.com/ECP-WarpX/WarpX/assets/126223397/18ef2521-a0cd-4d15-a392-38abfe84caae">

**Comment by ax3l** [comment]
Ok, then let us try the latest `OpenBLAS/0.3.21` to start with.

But let's build WarpX with the [above module environment first](https://github.com/ECP-WarpX/WarpX/issues/3938#issuecomment-1592094436) and let me know if that works, then we do advanced features.

```bash
cmake -S . -B build -DWarpX_COMPUTE=CUDA -DWarpX_QED_TABLE_GEN=ON -DWarpX_PSATD=ON -DWarpX_DIMS="1;2;3"
cmake --build build -j 12

cmake -S . -B build_rz -DWarpX_COMPUTE=CUDA -DWarpX_QED_TABLE_GEN=ON -DWarpX_PSATD=OFF -DWarpX_DIMS="RZ"
cmake --build build_rz -j 12
```

@floresv299 does this work? Can you post the build logs when running this? :)

**Comment by floresv299** [comment]
Dear Axel,

I've been trying to build Warpx for the past following days, apologies for the delay. I've gotten the following warnings / errors. I was hoping you would know how to go about them, thank you. I will attach the CMake error logs but essentially both RZ and 1/2/3D give the same error when building. 

these are the following warnings:

`warning #811-D: const variable "info" requires an initializer -- class "amrex::LPInfo" has no user-provided default constructor`
`warning #20014-D: calling a__host__ function from a __host__ __device__ function is not allowed`
`warning #177-D: function "amrex::<unnamed>::IFRTag::box" was declared but never referenced`
`tmpxft_002e0402_00000000-6_AMReX.compute_86.cudafe1.cpp:(.text+0x2418): failed to convert GOTPCREL relocation against '_ZN5amrex6system13error_handlerE'; relink with --no-relax`


Errors

for RZ

gmake[2]: *** [CMakeFiles/app_rz.dir/build.make:388: bin/warpx.rz.MPI.CUDA.DP.PDP.OPMD.QED] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1092: CMakeFiles/app_rz.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2

for 1/2/3D

collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/app_3d.dir/build.make:404: bin/warpx.3d.MPI.CUDA.DP.PDP.OPMD.PSATD.QED] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1369: CMakeFiles/app_3d.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
gmake: *** [Makefile:136: all] Error 2
[building_error_cart .txt](https://github.com/ECP-WarpX/WarpX/files/11780758/building_error_cart.txt)
[Building_errors_RZ.txt](https://github.com/ECP-WarpX/WarpX/files/11780760/Building_errors_RZ.txt)
[CMake_error_log_cart.txt](https://github.com/ECP-WarpX/WarpX/files/11780761/CMake_error_log_cart.txt)
[Cmake_Error_log_Rz.txt](https://github.com/ECP-WarpX/WarpX/files/11780762/Cmake_Error_log_Rz.txt)

**Comment by floresv299** [comment]
Aha,

I have fixed the error and successfully build the code. The problem was that I did not specified the target architecture. By adding the following (here Volta is the NVIDIA GPU generation)

`-DAMReX_CUDA_ARCH=Volta`

**Comment by ax3l** [comment]
Awesome, great news! I'll start a pull request (#4010) documenting the workflow and adding RZ geometry with PSATD field solver support, too.

Just checking, I think if you set
```bash
# optimize CUDA compilation for V100
export AMREX_CUDA_ARCH=7.0

# compiler environment hints
export CXX=$(which g++)
export CC=$(which gcc)
export FC=$(which gfortran)
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=${CXX}
```
and `rm -rf build` before rebuilding, then the variable `AMREX_CUDA_ARCH` should have the same effect for you. Can you confirm?

**Comment by ax3l** [comment]
Hi @floresv299 , can you post the output of 
```
scontrol show partition=gpu
scontrol show partition=free-gpu
scontrol show partition=gpu-debug
```
here? I am trying to document how many GPU nodes there are in HPC3 and [the respective homepage section](https://rcic.uci.edu/hpc3/slurm.html#memmap) is just shy of that detail for users.

**Comment by ax3l** [comment]
Hi @floresv299, can you please run

```bash
srun -p free-gpu --gres=gpu:V100:4 nvidia-smi
```

and post the output here? I cannot find how much memory each V100 GPU has and this will show us. What I found so far, there are four V100 GPUs per GPU node (nice!) :)

**Comment by floresv299** [comment]
> Hi @floresv299 , can you post the output of
> 
> ```
> scontrol show partition=gpu
> scontrol show partition=free-gpu
> scontrol show partition=gpu-debug
> ```
> 
> here? I am trying to document how many GPU nodes there are in HPC3 and [the respective homepage section](https://rcic.uci.edu/hpc3/slurm.html#memmap) is just shy of that detail for users.

Sure thing Axel, 
![GPU](https://github.com/ECP-WarpX/WarpX/assets/126223397/b753fb67-912b-4af9-8f28-660014a5bf04)
![freeGPU](https://github.com/ECP-WarpX/WarpX/assets/126223397/0b4d2748-7b35-414c-8127-a970de4291c2)

![GPU_debug](https://github.com/ECP-WarpX/WarpX/assets/126223397/fcefdc65-bc0c-4279-be44-a3ace2c05851)

**Comment by floresv299** [comment]
> Hi @floresv299, can you please run
> 
> ```shell
> srun -p free-gpu --gres=gpu:V100:4 nvidia-smi
> ```
> 
> and post the output here? I cannot find how much memory each V100 GPU has and this will show us. What I found so far, there are four V100 GPUs per GPU node (nice!) :)

In terms of GPU memory here is the output.
![GPUmem](https://github.com/ECP-WarpX/WarpX/assets/126223397/982c66ed-dd0b-48bf-8b97-f37db1f4de3f)

**Comment by ax3l** [comment]
Awesome, thanks a lot!

**Comment by floresv299** [comment]
> Awesome, great news! I'll start a pull request (#4010) documenting the workflow and adding RZ geometry with PSATD field solver support, too.
> 
> Just checking, I think if you set
> 
> ```shell
> # optimize CUDA compilation for V100
> export AMREX_CUDA_ARCH=7.0
> 
> # compiler environment hints
> export CXX=$(which g++)
> export CC=$(which gcc)
> export FC=$(which gfortran)
> export CUDACXX=$(which nvcc)
> export CUDAHOSTCXX=${CXX}
> ```
> 
> and `rm -rf build` before rebuilding, then the variable `AMREX_CUDA_ARCH` should have the same effect for you. Can you confirm?

Axel, I can confirm that after adding the above lines in your warpx.profile, the build is successful with no errors.

**Comment by ax3l** [comment]
Splendid, @floresv299!

To complete #4010, can you please try if the job script (`hpc3_gpu.sbatch`) I posted in there works? Can you post the output for `output.txt`, `WarpX.o*`, `WarpX.e*`?

**Comment by floresv299** [comment]
Axel,

I seem to have reached an error, OPAL ERROR: Unreachable in file pmix3x_client.c . I'm running on 1 GPU to begin. Here is what I get from the WarpX.e file, the rest are blank. 
[batcherror.txt](https://github.com/ECP-WarpX/WarpX/files/11792772/batcherror.txt)

**Comment by floresv299** [comment]
I'm not sure what the problem was but I managed to work around it. Looking through the HPC3 documentation for OpenMP Jobs, srun is not included. Here I attached the sbatch script and the output. The WarpX.e and Warpx.0 are empty. The outputs include a folder of diagnostics and a file called warpx_used_inputs. All the data seems to be here


[warpx_batch.txt](https://github.com/ECP-WarpX/WarpX/files/11793161/warpx_batch.txt)

[Warpx_out.txt](https://github.com/ECP-WarpX/WarpX/files/11793162/Warpx_out.txt)

**Comment by floresv299** [comment]
Here is the sbatch and output for multi MPI processes.

[Warpx_Sbatch_multi_MPI.txt](https://github.com/ECP-WarpX/WarpX/files/11793451/Warpx_Sbatch_multi_MPI.txt)

[Warpx_out_8_MPI_processe.txt](https://github.com/ECP-WarpX/WarpX/files/11793453/Warpx_out_8_MPI_processe.txt)

**Comment by ax3l** [comment]
Thanks for testing! :pray: 

I am looking at these docs:
- https://rcic.uci.edu/hpc3/slurm.html#_quick_start
- https://rcic.uci.edu/hpc3/examples.html

right now.

Good point, for this system, `mpirun -np $SLURM_NTASKS` seems to be the documented way for multi-GPU jobs, so now updated that. In your [test above](https://github.com/ECP-WarpX/WarpX/issues/3938#issuecomment-1597805456), I am a bit concerend that 8 MPI ranks were started but only 2 GPUs used. The ratio should be 1:1. I think the issue was that you only requested `#SBATCH --gres=gpu:V100:1`  and not `#SBATCH --gres=gpu:V100:4` per node, leaving 3 out of 4 GPUs unused for the two nodes requested (`#SBATCH --nodes=2`).

Please leave `#SBATCH --ntasks=...` unset - it will be automatically calculated, so that you can request your jobs in multiples of full nodes and only have to change a single line, `#SBATCH --nodes=...`, to scale the job size up and down :)

Can you try again to run first on one and then on two nodes, with the latest template that I pushed to #4010?
[hpc3_gpu.sbatch.txt](https://github.com/ECP-WarpX/WarpX/files/11793826/hpc3_gpu.sbatch.txt)

## Notes

In case you logged out of the system in between compiling and submission, it is important that you have again the `source $HOME/hpc3_gpu_warpx.profile` line activated, so all modules are correctly loaded for running.

This [comment in the docs](https://rcic.uci.edu/hpc3/examples.html#_special_case_openmpi_jobs) seems to suggest we actually might need to use some other flags, but let's try without first.

The [comment here](https://rcic.uci.edu/hpc3/examples.html#gpu-jobs) on GPU jobs does not apply to us:
> <div class="listingblock"><div class="content"><pre>#SBATCH <strong class="red">--gres=gpu:V100:1</strong>         # specify 1 GPU of type V100</pre>
</div>
</div>

> GPU number should be set to 1. Nearly 100% of applications on the cluster will use only 1 GPU. None of Perl, Python, R-based applications need multi-GPU. Very few applications can use multiple GPUs in P2P (peer-2-peer) mode. These applications need to be specially designed and compiled with very specific flags and options to be able to use multi-GPU acceleration. A few examples of applications that can do P2P are Amber and NAMD.

Like Amber and NAMD, WarpX does use all GPUs in parallel if we run 4 MPI ranks (or multiples thereof) on the system :)

If you want to run truly small simulations with WarpX, that only use one single GPU for a whole job, set this:
```
#SBATCH --gres=gpu:V100:1
#SBATCH --ntasks-per-node=1
```
This will leave both 3/4rds of GPUs and 3/4th of CPU cores of a node unused.

**Comment by ax3l** [comment]
I merged this first documentation workflow with #4010 in, so it can be rendered to https://warpx.readthedocs.io/en/latest/install/hpc/hpc3.html

We will do follow-up changes based on your tests and HPC support response for potential updates to the job script :)

**Comment by ax3l** [comment]
Hi @floresv299, how is it going?

Can you post again the output of a 1 and two node example run with the latest instructions to check all is read to go? :)

**Comment by floresv299** [comment]
Hi Axel, 
 
its going well, I hope you are doing good! Sure thing here are the outputs for runs with one and two nodes. 

`#SBATCH --nodes=1`
`#SBATCH --ntasks-per-node=4`
`#SBATCH --gres=gpu:V100:4`


`#SBATCH --nodes=2`
`#SBATCH --ntasks-per-node=4`
`#SBATCH --gres=gpu:V100:4`

The simulations run successfully however the following error arises: 
"Multiple GPUs are visible to each MPI rank, but the number of GPUs per socket or node has not been provided.
This may lead to incorrect or suboptimal rank-to-GPU mapping.!"

I have attached the files below. One quick note that I've noticed, I'm using Openpmd for analysis and h5 as backend. I was getting an error when I specified the backend as h5, turns out that when I was building the `build options` had `HDF5: OFF` . I tried to specify the path of the hdf5 along with its libraries but no luck. Then I changed the version from what we had `module load hdf5/1.14.0/gcc.11.2.0-openmpi.4.1.2` to `module load hdf5/1.13.1/gcc.11.2.0-openmpi.4.1.2`. That did the trick and I know get hf files for all geometries. 

[Warpx_run_1_node_output.txt](https://github.com/ECP-WarpX/WarpX/files/11824161/Warpx_run_1_node_output.txt)

[Warpx_run_1_node_err.txt](https://github.com/ECP-WarpX/WarpX/files/11824163/Warpx_run_1_node_err.txt)

[Warpx_run_2_node_output.txt](https://github.com/ECP-WarpX/WarpX/files/11824167/Warpx_run_2_node_output.txt)

[Warpx_run_2_node_err.txt](https://github.com/ECP-WarpX/WarpX/files/11824186/Warpx_run_2_node_err.txt)

**Comment by ax3l** [comment]
Hi @floresv299,

Thanks a lot for testing! Perfect, I reflected both your HDF5 module findings and an additional set of hints to address the GPU visibility issue you found in #4021.

Can you please update the WarpX profile with the changes in there?
Please re-login after the changes, activate the WarpX profile, remove the build directory and recompile WarpX to test the changes. Let me know what you find! :)

**Comment by floresv299** [comment]
Axel,

I made the changes to the Warpx profile, removed build directory, and recompiled Warpx. Unfortunately, the problem seems to persist. The same error is shown `Multiple GPUs are visible to each MPI rank, but the number of GPUs per socket or node has not been provided.This may lead to incorrect or suboptimal rank-to-GPU mapping.!`

**Comment by ax3l** [comment]
@floresv299,

Thanks for testing! Can you try again compiling with `cmake ... ... -DGPUS_PER_SOCKET=2 -DGPUS_PER_NODE=4` and see if that changes it?

**Comment by ax3l** [comment]
Hi @floresv299, I pushed another update to #4021 - can you try the new job script template and post the output? :)

**Comment by floresv299** [comment]
> @floresv299,
> 
> Thanks for testing! Can you try again compiling with `cmake ... ... -DGPUS_PER_SOCKET=2 -DGPUS_PER_NODE=4` and see if that changes it?

Axel, I just ran another simulation with these flags. I can confirm that this works ! No errors found, running on 
#SBATCH --nodes=1
#SBATCH -J WarpX
#SBATCH -p free-gpu
#SBATCH --ntasks-per-node=4
#SBATCH --gres=gpu:V100:4
#SBATCH --cpus-per-task=10

**Comment by floresv299** [comment]
Axel,

The code was now ran on 2 nodes. No errors were detected ! Here is the output. Thank you once again for all the help. A lab partner of mine was able to download and run the code with the current documentation. 
[8_MPI_pross_out.txt](https://github.com/ECP-WarpX/WarpX/files/11853762/8_MPI_pross_out.txt)

**Comment by ax3l** [comment]
This is awesome! Everything looks good to me!

On hint: I think the input file lacks a line `diag1.format = openpmd` and thus shows a warning about the backend line not being used.

We are all set, glad that you tested with a colleague that the instructions work for them as well 🎉

If you like, please feel free to close this issue and open new ones if you have any questions, suggestions or want to share your results in the future. Happy computing!!

**Comment by floresv299** [comment]
Thank you very much Axel, you and your team are amazing ! Really appreciate all the help.

**Comment by erny123** [comment]
Hi all, 

While running the install i get a similar error to [here](https://github.com/ECP-WarpX/WarpX/issues/3938#issuecomment-1595976745).

More specifically this line: 

`/data/homezvol0/ernestob/src/warpx/build/_deps/fetchedamrex-src/Src/AmrCore/AMReX_InterpFaceRegister.cpp(96): warning #177-D: function "amrex::<unnamed>::IFRTag::box" was declared but never referenced`

Tried Victor's solution: 
`-DAMReX_CUDA_ARCH=Volta`

But didn't work. 

[errorinstall.txt](https://github.com/ECP-WarpX/WarpX/files/12386407/errorinstall.txt)

**Comment by ax3l** [comment]
Hi, that is a warning in older CUDA compilers and safe to ignore.
Which CUDA Toolkit version did you use?

Besides the warning, your compile went through successfully? Or did you get the `failed to convert GOTPCREL relocation against` error?

**Comment by ax3l** [comment]
Looking at your error.txt, your error is:

nvcc error   : 'cicc' died due to signal 9 (Kill signal)

Please use a lower number in -j ... When you compile. You are exceeding the shared resources (mem and CPU) time of you shared login node and they thus terminate automatically some of your compile processes (Unix speak: kill your process)

**Comment by ax3l** [comment]
This is essentially #4181 and the fix is to lower -j #4182

Please let me know if -j12 is still to high and we should document to use -j8 or even 4.

Alternatively, we can compile on the compute nodes, which are not shared and all to yourself, by prefixing with runNode ...

**Comment by floresv299** [comment]
Hey Ernesto, 

can you please post your warpx.profile. Also are you building the RZ or the 1/2/3D code? I will attached my warpx.profile, it's a bit simpler but works. I also built the RZ and the 1/2/3D separately. Here are some commands that are missing in the current warpx.profile and in the cmake compile.

`export GPUS_PER_SOCKET=2`
`export GPUS_PER_NODE=4`

We need this because Slurm is not well configured to properly address GPU visibility. This will allow multi-GPU usage. When you build, make sure to include:

`-DGPUS_PER_SOCKET=${GPUS_PER_SOCKET}`
`-DGPUS_PER_NODE=${GPUS_PER_NODE}`


@ax3l  I don't think these commands are included in the current warpx.profile and in the cmake commands. Also I can confirm that -j 4 works !

[Compile Warpx.txt](https://github.com/ECP-WarpX/WarpX/files/12386462/Compile.Warpx.txt)
[HPC3_warpx_profile_victor.txt](https://github.com/ECP-WarpX/WarpX/files/12386464/HPC3_warpx_profile_victor.txt)

**Comment by floresv299** [comment]
One more thing, our cluster recently reinstalled all compute/gpu/login nodes with Enterprise Linux 8.8 and Slurm was updated to its latest release. This produced the following error when I tried running a simulation:
`---------------------------------------------------------------------------
Open MPI has detected that a parameter given to a command line
option does not match the expected format:
  Option: np
  Param:  /data/homezvol0/floresvr/warpx_source/build/bin/warpx.2d
This is frequently caused by omitting to provide the parameter
to an option that requires one. Please check the command line and try again.
----------------------------------------------------------------------------`

The work around that I found was replacing:
`mpirun -np ${SLURM_NTASKS}  \
  ${EXE} ${INPUTS}          \
  > output.txt`

with 

`mpirun -np ${SLURM_NTASKS_PER_NODE}  \
  ${EXE} ${INPUTS}          \
  > output.txt`

Not sure the reason for this.

**Comment by erny123** [comment]
@ax3l 

> Please use a lower number in -j ... When you compile. You are exceeding the shared resources (mem and CPU) time of you shared login node and they thus terminate automatically some of your compile processes (Unix speak: kill your process)

> Alternatively, we can compile on the compute nodes, which are not shared and all to yourself, by prefixing with runNode ...

That's exactly what I'm doing. I'm running the install on a script that requests 40 cores (whole node) but with the `-j 12` still in tact from the original documentation. 

```
#SBATCH -p free
#SBATCH --job-name=warpxinstall
#SBATCH --error
#SBATCH --nodes=1
#SBATCH --constraint=intel
#SBATCH --ntasks=40
#SBATCH --cpus-per-task=1

#SBATCH --mail-type=begin,end
#SBATCH --mail-user=ernestob@uci.edu
#SBATCH -o WarpXInstall.o%j
#SBATCH -e WarpXInstall.e%j

module purge

source $HOME/hpc3_gpu_warpx.profile
source $HOME/sw/hpc3/gpu/venvs/warpx-gpu/bin/activate
cd $HOME/src/warpx
rm -rf build

cmake -S . -B build -DCMAKE_VERBOSE_MAKEFILE=ON -DWarpX_COMPUTE=CUDA -DWarpX_PSATD=ON -DWarpX_QED_TABLE_GEN=ON -DWarpX_PYTHON=ON -DWarpX_DIMS="1;2;RZ;3"
cmake --build build -j 12
cmake --build build -j 12 --target pip_install

```

Perhaps I should be setting number of cpus over ntasks?

**Comment by erny123** [comment]
@floresv299 

`mpirun -np ${SLURM_NTASKS_PER_NODE}
${EXE} ${INPUTS} \`

Ya, I noticed that too in the script provided online when I tried running and was getting too many errors. 

Pretty sure the reason is: 
`#SBATCH --ntasks-per-node=4`

sets the variable: `${SLURM_NTASKS_PER_NODE} `

While: 
`#SBATCH --ntasks=4`

Sets the variable: 
`${SLURM_NTASKS}`

So this was already implemented in past SLURM versions but I think they tolerated the differences. In the new version that is no longer the case and they don't tolerate one as the other because you can easily have different amount of tasks per node. 

 As for the build, I'm trying to just build the 1D and 2D but it was throwing out too much errors that I had to go back to the original scripts documented online.

**Comment by erny123** [comment]
Attached is my warpx profile

[hpc3_gpu_warpx.profile.txt](https://github.com/ECP-WarpX/WarpX/files/12386545/hpc3_gpu_warpx.profile.txt)

**Comment by ax3l** [comment]
Can you try without the
GPUS_PER_SOCKET/NODE for the latest version? We updated this and do not use these options anymore.

**Comment by erny123** [comment]
Ok changing this in the compilation script:

```
#SBATCH --ntasks=40
#SBATCH --cpus-per-task=1
.
.
.
.
cmake -S . -B build -DCMAKE_VERBOSE_MAKEFILE=ON -DWarpX_COMPUTE=CUDA -DWarpX_PSATD=ON -DWarpX_QED_TABLE_GEN=ON -DWarpX_PYTHON=ON -DWarpX_DIMS="1;2" 
cmake --build build -j 32
cmake --build build -j 32 --target pip_install

```

to 

```
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=40
.
.
.
.
cmake -S . -B build -DCMAKE_VERBOSE_MAKEFILE=ON -DWarpX_COMPUTE=CUDA -DWarpX_PSATD=ON -DWarpX_QED_TABLE_GEN=ON -DWarpX_PYTHON=ON -DWarpX_DIMS="1;2" 
cmake --build build -j 32
cmake --build build -j 32 --target pip_install
```

Did not work. 

Maybe it's the
`#SBATCH --constraint=intel`

edit:
erasing the `intel` constraint did not work either. 

Alright it's time to stop being impatient and run `-j 4`

**Comment by erny123** [comment]
Same error using a compute node with 4 cpus to run: `-j 4`

@floresv299 have you tried installing on the compute nodes? Or have you only tried on the log-in nodes?

**Comment by erny123** [comment]
Using the log-in node and -j 4 looks like it fully compiled. Although, because it doesn't give an error output, I'm sure the same errors are still there, just not sure where these are outputted.

**Comment by erny123** [comment]
Looks like it's worked so far. 

@floresv299 is correct that the run script should be updated because the slurm variables: `${SLURM_NTASKS}`
only works when using:
`#SBATCH --ntasks=4`

Also, I think requesting the free gpu job needs to be at the top before requesting nodes. 
`#SBATCH -p free-gpu`

Here's one I used for 1DLWFA example 

```
#!/bin/bash -l

# Copyright 2023 The WarpX Community
#
# This file is part of WarpX.
#
# Authors: Axel Huebl, Victor Flores
# License: BSD-3-Clause-LBNL
#SBATCH -p free-gpu
#SBATCH --time=00:30:00
#SBATCH --nodes=1
#SBATCH -J WarpX
#SBATCH --ntasks=4
#SBATCH --gres=gpu:V100:4
#SBATCH --cpus-per-task=10
#SBATCH --mail-type=begin,end
#SBATCH --mail-user=ernestob@uci.edu
#SBATCH -o WarpX.o%j
#SBATCH -e WarpX.e%j

source /data/homezvol0/ernestob/hpc3_gpu_warpx.profile


# executable & inputs file or python interpreter & PICMI script here
EXE=python3
INPUTS=PICMI_inputs_1d.py

# OpenMP threads
export OMP_NUM_THREADS=${SLURM_CPUS_PER_TASK}
printf "number of tasks:\n"
printf "%s\n" "${SLURM_NTASKS}"
printf "number of threads:\n"
printf "%s\n" "$OMP_NUM_THREADS"

# run
mpirun -np ${SLURM_NTASKS} bash -c "export CUDA_VISIBLE_DEVICES=${SLURM_LOCALID}; ${EXE} ${INPUTS}"> output.txt
```
